### PR TITLE
feat: morph/react support aggregating multiple data keys

### PR DESCRIPTION
**Pull request by alex-vladut**
The data model was changed slightly to accommodate supporting a new `aggregate` function, so it basically looks like this:

```
[{
  "data": [
  {
    "context": "person",
    "path": "person.first_name",
    "format": { "formatIn": { "value": "trim", "source": "/Data/format.js"  }
  },
  {
    "context": "person",
    "path": "person.last_name"
  },
  ],
  "aggregate": { "value": "joinWithSpaces", "source": "/Data/aggregate.js" }
}]
```

The specified `aggregate` function will be applied on all the data keys in the order they were specified in e.g.

```
PersonInfo View
is together
  Info
  data person.first_name
  format /Data/format.js trim
  data person.last_name
  aggregate /Data/aggregate.js joinWithSpaces
  value <
```

Let me know if you think it make sense or if you have other suggestions on how we can improve it. Thanks!

### CODE REVIEW

File: parse/index.js [hunk 1]
```diff
@@ -359,12 +360,17 @@ export default ({
 
     let data = []
     let index = 0
+    let currentAggregateData = null
```
Comment (dariocravero) [inline parse/index.js:363]: Not sure how much this adds but what about naming it `currentDataGroup` to distinguish it from the `aggregate` keyword?

File: parse/index.js [hunk 2]
```diff
@@ -381,10 +387,26 @@ export default ({
...
+              !currentAggregateData.aggregate &&
+              currentAggregateData.data.length > 1
+            ) {
+              warnings.push({
+                type: `No aggregate function was provided, but ${currentAggregateData.data.length} data keys were found. Did you forget to specify an aggregate function?`,
+                line,
+                loc: block.loc,
+              })
+            }
+            currentAggregateData = null
```
Comment (dariocravero) [inline parse/index.js:403]: Not sure how feasible (and useful) it is given the code structure but before resetting this to null, do we want/need to set the end `loc` of `currentAggregateData` to match the last bit of data?

File: morph/react/block-add-data.js [hunk 3]
```diff
@@ -3,20 +3,44 @@ import toCamelCase from 'to-camel-case'
...
-    state.dataBlocks.push(
-      `let ${data.name} = fromData.useData({ viewPath, path: '${data.path}', `
-    )
-    maybeDataContext(data, state)
-    maybeDataFormat(data.format, state)
-    maybeDataValidate(data.validate, state)
-    state.dataBlocks.push('})')
-
-    state.use('ViewsUseData')
+  for (let aggregateData of node.data) {
```
Comment (dariocravero) [inline morph/react/block-add-data.js:6]: if we rename the var below to group, we should update it here too

File: morph/react/block-add-data.js [hunk 4]
```diff
@@ -3,20 +3,44 @@ import toCamelCase from 'to-camel-case'
...
+    }
+
+    if (aggregateData.aggregate) {
+      aggregateData.name = getAggregateDataVariableName(state)
+      let importName = getAggregateImportName(
+        aggregateData.aggregate.source,
+        state
+      )
+      state.dataBlocks.push(
+        `let ${aggregateData.name} = React.useMemo(() => ${importName}.${
```
Comment (dariocravero) [inline morph/react/block-add-data.js:30]: What do you think... Do we want to try first without memo or do you already see those being heavy enough to have it in to begin with?

File: parse/index.js (hunk 2 above)
Comment (alex-vladut) [reply parse/index.js:403]: yes, I can do that. didn't add it as for now only the start is used

File: morph/react/block-add-data.js (hunk 4 above)
Comment (alex-vladut) [reply morph/react/block-add-data.js:30]: not sure, probably the aggregate functions should be fast enough. I could get rid of the useMemo for now and we can reevaluate if it is needed later

Comment (dariocravero) [reply morph/react/block-add-data.js:30]: cool, maybe leave the code that generates the memo commented out so we have it handy for later just in case?

Comment (alex-vladut) [reply morph/react/block-add-data.js]: do you think this is necessary to have the useMemo commented out? the logic is quite simple so I think we can easily put it back when needed

Comment (dariocravero) [reply morph/react/block-add-data.js]: entirely up to you :D

**Review by dariocravero — APPROVED** [review]

**Comment by dariocravero** [comment]
nice one! :)

**Pull request merged**